src/ScriptVerification/StructuredContracts.lagda

```diff
@@ -0,0 +1,52 @@
+\section{Structured Contracts}
+
+\begin{code}[hide]
+{-# OPTIONS --safe #-}
+
+open import Algebra
+open import Data.Nat.Properties using (+-0-monoid)
+
+open import Ledger.Prelude; open Equivalence
+open import Ledger.Transaction
+open import Ledger.Abstract
+open import Ledger.TokenAlgebra 
+open import Ledger.TokenAlgebra.ValueSet
+
+
+module ScriptVerification.StructuredContracts
+  (txs : _) (open TransactionStructure txs)
+  (abs : AbstractFunctions txs) (open AbstractFunctions abs)
+  where
+
+open import Ledger.Ledger txs abs
+open import Ledger.ScriptValidation txs abs
+open import Ledger.Utxo txs abs
+\end{code}
+-- open import Prelude hiding (id; _~ˢ_; _~ᵉ_)
+
+\begin{code}
+SSRel : Type → Type → Type → Type₁
+SSRel Env State Input = Env → State → Input → State → Type
+
+record StrucSimulation S I (_⊢_⇀⦇_,STRUC⦈_ : SSRel ⊤ S I) : Type₁ where
+  field
+    πˢ : LState → Maybe S
+    πⁱ : TxInfo → I
+    extraAssmp : LEnv → LState → Tx → Type -- extra assumptions about ledger state, env, and transaction that 
+
+  _~ˢ_ : LState → S → Type -- R
+  u ~ˢ s = πˢ u ≡ just s
+
+    -- only for PlutusV3
+  _~ᵉ_ : LEnv × Tx × LState → ⊤ × I → Type
+  _~ᵉ_ = λ (le , tx , u) (tt , i) → πⁱ (txInfo PlutusV3 (le .LEnv.pparams) (u .LState.utxoSt .UTxOState.utxo) tx)  ≡ i
```

Comment:
```suggestion
  (le , tx , u) ~ᵉ (tt , i) = πⁱ (txInfo PlutusV3 (le .LEnv.pparams) (u .LState.utxoSt .UTxOState.utxo) tx) ≡ i
```

Reply:
hmm the compiler does not like this
"Not a valid let-declaration when scope checking the declaration"

Reply:
Ah, right, this is forbidden inside of a record. It's a shame because it would look a bit better and does the same thing, but that's just a limitation here.